server/src/com/thoughtworks/go/server/service/AgentConfigService.java

```diff
@@ -47,45 +68,446 @@ public void register(AgentChangeListener agentChangeListener) {
-        goConfigService.saveOrUpdateAgent(agentInstance);
+    public void saveOrUpdateAgent(AgentInstance agentInstance, Username currentUser) {
+        AgentConfig agentConfig = agentInstance.agentConfig();
+        if (goConfigService.hasAgent(agentConfig.getUuid())) {
+            this.updateAgentApprovalStatus(agentConfig.getUuid(), agentConfig.isDisabled(), currentUser);
+        } else {
+            this.addAgent(agentConfig, currentUser);
+        }
     }
 
+    @Deprecated
     public void approvePendingAgent(AgentInstance agentInstance) {
-        goConfigService.approvePendingAgent(agentInstance);
+        agentInstance.enable();
+        if (goConfigService.hasAgent(agentInstance.getUuid())) {
+            LOGGER.warn("Registered agent with the same uuid [" + agentInstance + "] already approved.");
+        } else {
+            updateAgent(createAddAgentCommand(agentInstance.agentConfig()), agentInstance.getUuid(), new HttpOperationResult(), Username.ANONYMOUS);
+        }
     }
 
-    // For tests
-    public void updateAgentApprovalStatus(String uuid, boolean isDenied) {
-        goConfigService.updateAgentApprovalStatus(uuid, isDenied);
+    protected static UpdateConfigCommand createAddAgentCommand(final AgentConfig agentConfig) {
+        return new AddAgentCommand(agentConfig);
     }
 
+    /**
+     * @understands how to add an agent to the config file
+     */
+    private static class AddAgentCommand implements UpdateConfigCommand {
+        private final AgentConfig agentConfig;
+
+        public AddAgentCommand(AgentConfig agentConfig) {
+            this.agentConfig = agentConfig;
+        }
+
+        public CruiseConfig update(CruiseConfig cruiseConfig) throws Exception {
+            cruiseConfig.agents().add(agentConfig);
+            return cruiseConfig;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+
+            AddAgentCommand that = (AddAgentCommand) o;
+
+            if (agentConfig != null ? !agentConfig.equals(that.agentConfig) : that.agentConfig != null) {
+                return false;
+            }
+
+            return true;
+        }
+
+        @Override
+        public int hashCode() {
+            return agentConfig != null ? agentConfig.hashCode() : 0;
+        }
+
+        @Override
+        public String toString() {
+            return "AddAgentcommand{" +
+                    "agentConfig=" + agentConfig +
+                    '}';
+        }
 
-    public void modifyResources(AgentInstance[] agentInstances, List<TriStateSelection> selections) {
-        goConfigService.modifyResources(agentInstances, selections);
     }
 
+    @Deprecated
+    public void updateAgentResources(final String uuid, final Resources resources) {
```

Comment:
This does not seem to be used, outside of tests, can we delete this method (or move it to a test helper)?